packages/query-core/src/hydration.ts

```diff
@@ -178,9 +179,17 @@ export function hydrate(
     }
 
     if (promise) {
+      const { transformPromise } = client.getDefaultOptions().hydrate ?? {}
+
+      // Note: `Promise.resolve` required cause
+      // RSC transformed promises are not thenable
+      const initialPromise = transformPromise
+        ? transformPromise(Promise.resolve(promise))
+        : promise
+
```

Comment:
not a bit fan of using `{}` as a fallback because it might widen the type to `| {}`. We could just keep it undefined and then use optional chaining as well:

```suggestion
      const transformPromise =
        client.getDefaultOptions().hydrate?.transformPromise

      // Note: `Promise.resolve` required cause
      // RSC transformed promises are not thenable
      const initialPromise =
        transformPromise?.(Promise.resolve(promise)) ?? promise

```